### PR TITLE
User version fixes and refactoring.

### PR DESCRIPTION
**Pull request by justinlittman**
## Why was this change made? 🤔
To allow user versioning to work correctly.

## How was this change tested? 🤨

⚡ ⚠ If this change has cross service impact, including data writes to shared file systems, ***run [integration tests](https://github.com/sul-dlss/infrastructure-integration-test)*** and/or test in [stage|qa] environment, in addition to specs. ⚡

Unit, QA

### CODE REVIEW

File: app/controllers/versions_controller.rb [hunk 1]
```diff
@@ -88,7 +88,8 @@ def close_params
     new_params = params.permit(
       :description,
       :start_accession,
-      :user_name
+      :user_name,
+      :user_versions
```
Comment (justinlittman) [inline app/controllers/versions_controller.rb:92]: The param wasn't permitted.

File: app/models/repository_object.rb [hunk 2]
```diff
@@ -12,6 +12,7 @@ class VersionAlreadyOpened < StandardError; end
   class VersionNotOpened < StandardError; end
 
   has_many :versions, -> { order(version: :asc) }, class_name: 'RepositoryObjectVersion', dependent: :destroy, inverse_of: 'repository_object'
+  has_many :user_versions, through: :versions
```
Comment (justinlittman) [inline app/models/repository_object.rb:15]: Adding this makes some operations below easier.

File: app/services/user_version_service.rb [hunk 3]
```diff
@@ -15,7 +15,7 @@ def self.create(druid:, version:)
     raise(UserVersioningError, 'RepositoryObjectVersion not closed') unless repository_object_version.closed?
 
     # Get the next increment of the user version (or 1 if this is the first user version)
-    next_user_version = repository_object_version.user_versions.maximum(:version)&.next || 1
+    next_user_version = repository_object_version.repository_object.user_versions.maximum(:version)&.next || 1
```
Comment (justinlittman) [inline app/services/user_version_service.rb:18]: You want the max of all user versions, not just for this this repository object version.

File: app/services/version_service.rb [hunk 4]
```diff
@@ -38,12 +38,12 @@ def self.can_open?(druid:, version:, assume_accessioned: false)
   # @param [String] description describes the version change
   # @param [String] user_name add username to the events datastream
   # @param [Boolean] start_accession (true) set to true if you want accessioning to start, false otherwise
-  # @param [String] user_version (none) one of none, new, update
-  def self.close(druid:, version:, description: nil, user_name: nil, start_accession: true, user_version: 'none')
+  # @param [String] user_versions (none) one of none, new, update
```
Comment (justinlittman) [inline app/services/version_service.rb:41]: Align the parameter with openapi, which specifies "user_versions".

File: app/services/version_service.rb [hunk 5]
```diff
@@ -149,13 +147,7 @@ def close(description:, user_name:, start_accession: true, user_version: 'none')
...
     EventFactory.create(druid:, event_type: 'version_close', data: { who: user_name, version: version.to_s })
 
-    UserVersionService.create(druid:, version: repository_object.last_closed_version.version) if user_version == 'new'
-
-    return unless user_version == 'update' && current_closed_version
-
-    current_closed_version.user_versions.each do |closed_user_version|
-      UserVersionService.move(druid:, version: repository_object.last_closed_version.version, user_version: closed_user_version.version)
-    end
+    update_user_version(user_versions:, repository_object:)
```
Comment (justinlittman) [inline app/services/version_service.rb:150]: Moved the logic into its own method.

File: app/services/version_service.rb [hunk 6]
```diff
@@ -216,4 +208,15 @@ def ensure_closeable!
     raise VersionService::VersioningError, "Trying to close version #{version} on #{druid} which has active assemblyWF" if assembling?
     raise VersionService::VersioningError, "accessionWF already created for versioned object #{druid}" if accessioning?
   end
+
+  def update_user_version(user_versions:, repository_object:)
+    return if user_versions == 'none'
+    return UserVersionService.create(druid:, version: repository_object.last_closed_version.version) if user_versions == 'new'
+
+    # If called with 'update', but there are no user versions to update, create a new user version.
+    return UserVersionService.create(druid:, version: repository_object.last_closed_version.version) if repository_object.user_versions.empty?
```
Comment (justinlittman) [inline app/services/version_service.rb:217]: This is new behavior.

File: app/services/version_service.rb [hunk 7]
```diff
@@ -216,4 +208,15 @@ def ensure_closeable!
...
+
+    max_user_version = repository_object.user_versions.maximum(:version)
```
Comment (justinlittman) [inline app/services/version_service.rb:219]: Changed from just looking at the last repository object version.

File: app/services/version_service.rb [hunk 8]
```diff
@@ -216,4 +208,15 @@ def ensure_closeable!
     raise VersionService::VersioningError, "Trying to close version #{version} on #{druid} which has active assemblyWF" if assembling?
     raise VersionService::VersioningError, "accessionWF already created for versioned object #{druid}" if accessioning?
   end
+
+  def update_user_version(user_versions:, repository_object:)
+    return if user_versions == 'none'
+    return UserVersionService.create(druid:, version: repository_object.last_closed_version.version) if user_versions == 'new'
+
+    # If called with 'update', but there are no user versions to update, create a new user version.
```
Comment (mjgiarlo) [inline app/services/version_service.rb:216]: Appreciate this comment!

**Review by mjgiarlo — APPROVED** [review]

**Pull request merged**